applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h

```diff
@@ -110,40 +112,243 @@ class MapperVertexMorphingMatrixFree : public Mapper
-    void MapToDesignSpace( const Variable<array_3d> &rNodalVariable, const Variable<array_3d> &rNodalVariableInDesignSpace ) override
+    void Map( const Variable<array_3d> &rVariable, const Variable<array_3d> &rMappedVariable ) override
     {
+        if (mIsMappingInitialized == false)
+            Initialize();
+
         BuiltinTimer mapping_time;
-        std::cout << "\n> Starting to map " << rNodalVariable.Name() << " to design space..." << std::endl;
+        std::cout << "\n> Starting mapping of " << rVariable.Name() << "..." << std::endl;
+
+        // Prepare vectors for mapping
+        mXValuesDestination.clear();
+        mYValuesDestination.clear();
+        mZValuesDestination.clear();
+
+        // Perform mapping
+        for(auto& node_i : mrDestinationMdpa.Nodes())
+        {
+            NodeVector neighbor_nodes(mMaxNumberOfNeighbors);
+            std::vector<double> resulting_squared_distances(mMaxNumberOfNeighbors);
+            unsigned int number_of_neighbors = mpSearchTree->SearchInRadius( node_i,
+                                                                             mFilterRadius,
+                                                                             neighbor_nodes.begin(),
+                                                                             resulting_squared_distances.begin(),
+                                                                             mMaxNumberOfNeighbors );
+
+            ThrowWarningIfNumberOfNeighborsExceedsLimit(node_i, number_of_neighbors);
+
+            std::vector<double> list_of_weights( number_of_neighbors, 0.0 );
+            double sum_of_weights = 0.0;
+            ComputeWeightForAllNeighbors( node_i, neighbor_nodes, number_of_neighbors, list_of_weights, sum_of_weights );
+
+            int node_i_mapping_id = node_i.GetValue(MAPPING_ID);
+            for(unsigned int neighbor_itr = 0 ; neighbor_itr<number_of_neighbors ; neighbor_itr++)
+            {
+                double weight = list_of_weights[neighbor_itr] / sum_of_weights;
+
+                ModelPart::NodeType& node_j = *neighbor_nodes[neighbor_itr];
+                array_3d& nodal_variable = node_j.FastGetSolutionStepValue(rVariable);
+
+                mXValuesDestination[node_i_mapping_id] += weight*nodal_variable[0];
+                mYValuesDestination[node_i_mapping_id] += weight*nodal_variable[1];
+                mZValuesDestination[node_i_mapping_id] += weight*nodal_variable[2];
+            }
+        }
+
+        // Assign results to nodal variable
+        for(auto& node_i : mrDestinationMdpa.Nodes())
+        {
+            int i = node_i.GetValue(MAPPING_ID);
 
-        CreateSearchTreeIfNecessary();
-        ClearVectorsForMappingToDesignSpace();
-        MapVariableComponentwiseToDesignSpace( rNodalVariable );
-        AssignResultingDesignVectorsToNodalVariable( rNodalVariableInDesignSpace );
+            Vector node_vector = ZeroVector(3);
```

Comment:
```suggestion
            Vector& r_node_vector = node_i.FastGetSolutionStepValue(rMappedVariable);
            r_node_vector(0) = mXValuesDestination[i];
            r_node_vector(1) = mYValuesDestination[i];
            r_node_vector(2) = mZValuesDestination[i];
```

Probably also applicable for the other mappers.